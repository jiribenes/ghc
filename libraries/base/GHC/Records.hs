{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015-2016
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- This module defines the 'HasField' class used by the
-- @OverloadedRecordFields@ extension.  See the
-- <https://gitlab.haskell.org/ghc/ghc/wikis/records/overloaded-record-fields
-- wiki page> for more details.
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       , getField
       , setField
       ) where

-- | Constraint representing the fact that the field @x@ can be get and set on
--   the record type @r@ and has field type @a@.  This constraint will be solved
--   automatically, but manual instances may be provided as well.
--
--   The function should satisfy the invariant:
--
-- > uncurry ($) (hasField @x r) == r

--   HasField :: forall {k}. k -> * -> * -> Constraint
--   getField :: forall {k} (x::k) r a. HasField x r a => r -> a
-- NB: The {k} means that k is an 'inferred' type variable, and
--     hence not provided in visible type applications.  Thus you
--     say     getField @"foo"
--     not     getField @Symbol @"foo"
class HasField x r a | x r -> a where
  -- | Function to get and set a field in a record.
  hasField :: r -> (a -> r, a)

-- | Selector function to extract the field from the record.
getField :: forall x r a . HasField x r a => r -> a
getField = snd . hasField @x

-- | Update function to set a new value of the field on the record.
setField :: forall x r a . HasField x r a => r -> a -> r
setField = fst . hasField @x
