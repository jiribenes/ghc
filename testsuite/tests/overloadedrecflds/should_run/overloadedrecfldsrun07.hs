{-# LANGUAGE DataKinds
           , PolyKinds
           , FlexibleContexts
           , FlexibleInstances
           , GADTs
           , MultiParamTypeClasses
           , OverloadedLabels
           , ScopedTypeVariables
           , StandaloneDeriving
           , TypeApplications
           , TypeOperators
           , UndecidableInstances
  #-}

import GHC.OverloadedLabels
import GHC.Records
import GHC.TypeLits
import Data.Kind
import Data.Proxy

data Label (x :: Symbol) = Label

instance KnownSymbol x => Show (Label x) where
  show _ = "#" ++ symbolVal (Proxy @x)

data Labelled x a = Label x := a
  deriving Show

data Rec :: forall k. [(k, Type)] -> Type where
  Nil  :: Rec '[]
  (:>) :: Labelled x a -> Rec xs -> Rec ('(x, a) ': xs)
instance Show (Rec '[]) where
  show Nil = "Nil"
deriving instance (KnownSymbol x, Show a, Show (Rec xs)) => Show (Rec ('(x, a) ': xs))
infixr 5 :>

instance {-# OVERLAPS #-} a ~ b => HasField foo (Rec ('(foo, a) ': xs)) b where
  hasField ((l := v) :> xs) = (\ v' -> (l := v') :> xs, v)

instance HasField foo (Rec xs) b => HasField foo (Rec ('(bar, a) ': xs)) b where
  hasField (x :> vs) = (\ v -> x :> setField @foo vs v, getField @foo vs)

instance y ~ x => IsLabel y (Label x) where
  fromLabel = Label

instance HasField x r a => IsLabel x (r -> a) where
  fromLabel = getField @x

x :: Rec '[ '("foo", Int), '("bar", Bool)]
x = #foo := 42 :> #bar := True :> Nil

y = #bar := 'x' :> undefined

main = do print (#foo x)
          print (#bar x)
          print (#bar y)
          print (setField @"foo" x 11)
