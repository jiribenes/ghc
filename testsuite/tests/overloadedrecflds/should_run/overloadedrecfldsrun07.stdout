42
True
'x'
#foo := 11 :> (#bar := True :> Nil)
